File: hotalg/wish/L19.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

"""
给你一个链表，删除链表的倒数第 n 个结点，并且返回链表的头结点。
"""
class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        dummy = ListNode()
        dummy.next = head
        p = dummy
        q = dummy
        for i in range(0, n+1):
            q = q.next
        while q != None:
            p = p.next
            q = q.next

        p.next = p.next.next
        return dummy.next

File: algorithm/src/main/java/sulqn/L19.java
package sulqn;

public class L19 {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode();
        dummy.next = head;
        ListNode a = dummy;
        ListNode b = head;
        ListNode c = head;
        for (int i = 0; i < n; i++) {
            c = c.next;
        }

        while (c != null) {
            a = a.next;
            b = b.next;
            c = c.next;
        }
        a.next = b.next;
        return dummy.next;
    }
}
